File: SPRING/스프링 입문 - 코드로 배우는/Ch7.AOP/hello-spring/src/main/java/hello/hellospring/service/MemberService.java
package hello.hellospring.service;

import hello.hellospring.domain.Member;
import hello.hellospring.repository.MemberRepository;
import hello.hellospring.repository.MemoryMemberRepository;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Transactional
public class MemberService {

    private final MemberRepository memberRepository;

    public MemberService(MemberRepository memberRepository){
        this.memberRepository = memberRepository; // 외부에서 넣어 주도록
    }

    /**
     * 회원 가입
     */

    public Long join(Member member) {
        // 일일이 시간 체크 방법
//
//        long start = System.currentTimeMillis();
//
//        try {
//            //이름 중복 검사
//            validateDuplicateMember(member);
//            memberRepository.save(member);
//            return member.getId();
//
//        } finally {
//            long finish = System.currentTimeMillis();
//            long timeMs = finish - start;
//            System.out.println("join = " + timeMs + "ms");
//        }

        //이름 중복 검사
        validateDuplicateMember(member);
        memberRepository.save(member);
        return member.getId();

    }

    private void validateDuplicateMember(Member member) {
        memberRepository.findByName(member.getName())
                .ifPresent(m -> {
                    throw new IllegalStateException("이미 존재하는 회원입니다.");
                } );
    }

    /**
     * 전체 회원 조회
     */
    public List<Member> findMembers() {
        return memberRepository.findAll();
    }

    public Optional<Member> findOne(Long memberId){
        return  memberRepository.findById(memberId);
    }
}

File: SPRING/simple-blog/src/main/kotlin/com/yuseogi/simpleblog/service/MemberService.kt
package com.yuseogi.simpleblog.service

import com.yuseogi.simpleblog.domain.member.*
import com.yuseogi.simpleblog.exception.MemberNotFoundException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class MemberService(
    private val memberRepository: MemberRepository
) {

    @Transactional(readOnly = true)
    fun findAll(pageable: Pageable): Page<MemberRes> =
        memberRepository.findMembers(pageable).map {
            it.toDto()
        }

    @Transactional(readOnly = true)
    fun findById(id: Long): MemberRes {
        return memberRepository.findById(id).orElseThrow {
            throw MemberNotFoundException(id.toString())
        }.toDto()
    }

    @Transactional
    fun delete(id: Long) {
        return memberRepository.deleteById(id)
    }

}